File: 算法竞赛进阶指南/0x00基本算法/AcWing100.py
N = 100006
a, b = [0]*N, [0]*N

def main():
    n = int(input())
    for i in range(1, n+1): a[i] = int(input())
    b[n+1] = 0; b[1] = a[1]
    for i in range(2, n+1): b[i] = a[i] - a[i-1]
    p, q = 0, 0
    for i in range(2, n+1):
        if b[i] > 0: p += b[i]
        elif b[i] < 0: q -= b[i]
    print(max(p, q))
    print(abs(p - q) + 1)

main()

File: 0x00基本算法/0x03前缀和与差分/AcWing100.cpp
#include <iostream>
#define ll long long
using namespace std;
const int N = 100006;
int a[N], b[N]; // b是a上的差分数列，问题转化为把b2,..,bn变全0

int main() {
  int n;
  cin >> n;
  for (int i=1; i<=n; i++) cin >> a[i];
  b[n+1] = 0, b[1] = a[1];
  for (int i=2; i<=n; i++) b[i] = a[i] - a[i-1];
  // 这里因为b[i]可能是0x7fffffff，两个数相加会溢出，所以用ll
  ll p = 0, q = 0;
  for (int i=2; i<=n; i++)
    if (b[i] > 0) p += b[i];
    else if (b[i] < 0) q -= b[i];
  cout << max(p, q) << endl << abs(p - q) + 1;
  return 0;
}